### ISSUE TITLE
Is the PhaseNet architecture the same with the one in the original paper (Zhu et al, 2019) ?

### ISSUE THREAD

**Issue by zhong-yy** [seisbench/seisbench#151]
This is the PhaseNet architecture from Zhu, W., & Beroza, G. C. (Geophysical Journal International, 2019, 216(1), 261-273.)
![Screenshot from 2022-12-13 21-34-21](https://user-images.githubusercontent.com/38816011/207336595-63c4d1a3-1a7b-46ec-aae5-7156be755acf.png)
Here at each depth, there are two layers. Following each down-sampling/up-sampling convolutional layer, there seems to be a 1x1 convolution layer that increases/shrinks the number of channels.

The following code is from the seisbench package (`seisbench/models/phasenet.py`). There is only one down-sampling/up-sampling layer at each depth. I am wondering if this implementation is equivalent to the above figure.
```python
    def __init__(
        self, in_channels=3, classes=3, phases="NPS", sampling_rate=100, **kwargs
    ):
        ...
        self.in_channels = in_channels
        self.classes = classes
        self.kernel_size = 7
        self.stride = 4
        self.activation = torch.relu

        self.inc = nn.Conv1d(self.in_channels, 8, 1)
        self.in_bn = nn.BatchNorm1d(8)

        self.conv1 = Conv1dSame(8, 11, self.kernel_size, self.stride)
        self.bnd1 = nn.BatchNorm1d(11)

        self.conv2 = Conv1dSame(11, 16, self.kernel_size, self.stride)
        self.bnd2 = nn.BatchNorm1d(16)

        self.conv3 = Conv1dSame(16, 22, self.kernel_size, self.stride)
        self.bnd3 = nn.BatchNorm1d(22)

        self.conv4 = Conv1dSame(22, 32, self.kernel_size, self.stride)
        self.bnd4 = nn.BatchNorm1d(32)

        self.up1 = nn.ConvTranspose1d(
            32, 22, self.kernel_size, self.stride, padding=self.conv4.padding
        )
        self.bnu1 = nn.BatchNorm1d(22)

        self.up2 = nn.ConvTranspose1d(
            44,
            16,
            self.kernel_size,
            self.stride,
            padding=self.conv3.padding,
            output_padding=1,
        )
        self.bnu2 = nn.BatchNorm1d(16)

        self.up3 = nn.ConvTranspose1d(
            32, 11, self.kernel_size, self.stride, padding=self.conv2.padding
        )
        self.bnu3 = nn.BatchNorm1d(11)

        self.up4 = nn.ConvTranspose1d(22, 8, self.kernel_size, self.stride, padding=3)
        self.bnu4 = nn.BatchNorm1d(8)

        self.out = nn.ConvTranspose1d(16, self.classes, 1)
        self.softmax = torch.nn.Softmax(dim=1)

    def forward(self, x):
        x_in = self.activation(self.in_bn(self.inc(x)))

        x1 = self.activation(self.bnd1(self.conv1(x_in)))
        x2 = self.activation(self.bnd2(self.conv2(x1)))
        x3 = self.activation(self.bnd3(self.conv3(x2)))
        x4 = self.activation(self.bnd4(self.conv4(x3)))

        x = torch.cat([self.activation(self.bnu1(self.up1(x4))), x3], dim=1)
        x = torch.cat([self.activation(self.bnu2(self.up2(x))), x2], dim=1)
        x = torch.cat([self.activation(self.bnu3(self.up3(x))), x1], dim=1)
        x = torch.cat([self.activation(self.bnu4(self.up4(x))), x_in], dim=1)

        x = self.out(x)
        x = self.softmax(x)

        return x
```

**Comment by JUNZHU-SEIS** [comment]
Besides, the architecture of original phasenet has different numbers of filters from the original paper too.
For example, the filter number of the second downsampling layer is 11,11,16 in the original paper, while in the source code of phasenet repository. The filter number is 8,8,16.
[Here](https://github.com/AI4EPS/PhaseNet/blob/164b17eee69941515c0fa397ea0cd56f111170f7/phasenet/model.py#L183) are more details

**Comment by JUNZHU-SEIS** [comment]
Dropout is also implemented in the original phasenet.

**Comment by JUNZHU-SEIS** [comment]
Actually I have trained many phasenet models through the seisbench API, all of them seem have achieved high recall but low precision, especially for continuous data. So I design a torch version PhaseNet model which has an architecture same as the original one, maybe it can help us reduce false picks.

**Comment by yetinam** [comment]
Hey @JUNZHU-SEIS,

thanks for the detailed analysis and the PR with the code. I'll only get to address the issue in early January but this will without question be a major help.
